internal/resource/registry.go

```diff
@@ -0,0 +1,58 @@
+package resource
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+type Registry interface {
+	// Register the given resource type and its hooks.
+	Register(reg Registration)
+
+	// Resolve the given resource type and its hooks.
+	Resolve(typ *pbresource.Type) (reg Registration, ok bool)
+}
+
+type Registration struct {
```

Comment:
We can definitely tackle this in a follow-up, but I've just been reflecting that we probably want to keep a reference to the protobuf message type on here (so that we can validate it on writes).

Protobufs has some prior art for this in its own [global type registry](https://github.com/golang/protobuf/blob/5d5e8c018a13017f9d5b8bf4fad64aaa42a87308/proto/registry.go).

Reply:
Added TODO in https://github.com/hashicorp/consul/pull/16622/commits/83ff27e7618383ccfaf1c0b54fb3d5d52bf1eeb8